src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,840 @@
+      _openslide_hash_data(quickhash1, &tag, sizeof(tag));
+      // need to get the vr class
+      //_openslide_hash_data(quickhash1, data, datalen);
+    }
+  }
+}
+
+static bool dicom_open(openslide_t *osr,
+                       const char *filename,
+                       struct _openslide_tifflike *tl G_GNUC_UNUSED,
+                       struct _openslide_hash *quickhash1,
+                       GError **err) {
+  g_autofree char *dirname = g_path_get_dirname(filename);
+
+  g_autoptr(_openslide_dir) dir = _openslide_dir_open(dirname, err);
+  if (!dir) {
+    return false;
+  }
+
+  g_autoptr(GHashTable) dicom_file_hash =
+    g_hash_table_new_full(g_str_hash,
+                          g_str_equal,
+                          g_free,
+                          (GDestroyNotify) dicom_file_destroy);
+
+  // open the passed-in file and get the slide-id
+  struct dicom_file *start = dicom_file_new(filename, err);
+  if (!start) {
+    return false;
+  }
+  g_hash_table_insert(dicom_file_hash, g_strdup(filename), start);
+
+  const char *slide_id;
+  if (!read_whole_file(start) ||
+    !get_tag_str(start->metadata, SeriesInstanceUID, 0, &slide_id)) {
+    return false;
+  }
+
+  // open all DICOM files that look like parts of this slide image and
+  // get the file metadata
+  const char *name;
+  while ((name = _openslide_dir_next(dir))) {
+    g_autofree char *path = g_build_path("/", dirname, name, NULL);
+
+    // no need to open the start file again
+    if (g_str_equal(path, filename)) {
+      continue;
+    }
+
+    g_debug("trying to open: %s ...", path);
+    GError *local_error = NULL;
+    g_autoptr(dicom_file) f = dicom_file_new(path, &local_error);
+    if (!f) {
+      g_debug("open failed: %s", local_error->message);
+      g_error_free(local_error);
+      continue;
+    }
+
+    const char *this_slide_id;
+    if (!read_whole_file(f) ||
+      !get_tag_str(f->metadata, SeriesInstanceUID, 0, &this_slide_id) ||
+      !g_str_equal(this_slide_id, slide_id)) {
+      continue;
+    }
+
+    g_hash_table_insert(dicom_file_hash,
+      g_steal_pointer(&path), g_steal_pointer(&f));
+  }
+
+  g_debug("found WSI DICOM files:");
+  g_hash_table_foreach(dicom_file_hash, print_file_from_hash, NULL);
+
+  // pull out the subset of these files that look like pyramid levels
+  g_autoptr(GPtrArray) level_array =
+    g_ptr_array_new_full(g_hash_table_size(dicom_file_hash),
+		         (GDestroyNotify) level_destroy);
+
+  g_hash_table_foreach_steal(dicom_file_hash, find_levels, level_array);
+
+  // make the rendering grid
+  for (guint i = 0; i < level_array->len; i++) {
+    struct dicom_level *l = (struct dicom_level *) level_array->pdata[i];
+
+    l->grid = _openslide_grid_create_simple(osr,
+					    l->tiles_across,
+					    l->tiles_down,
+					    l->base.tile_w,
+					    l->base.tile_h,
+					    read_tile);
+  }
+
+  // now sort levels by base.w to make the level array
+  g_ptr_array_sort(level_array, compare_level_width);
```

Comment:
It's possible that we could end up with zero levels here.  We should detect that case and fail.

Reply:
OK, added an error.